### ISSUE TITLE
Modifiers on Sequences

### ISSUE THREAD

**Issue by cknadler** [cknadler/marktab-c#25]
Should we allow modifiers on sequences? Currently, my implmentation isn't. 

Examples:

```
[1:4 1:3 1:2]*
```

```
riff: [1:4 1:3 1:2]

riff*
```

/cc @russplaysguitar

**Comment by russplaysguitar** [comment]
Oooh yeah, I could see that being really useful for chord slides:

```
(6:1 5:3 4:3) / (6:3 5:5 4:5)
```

```
F: (6:1 5:3 4:3)
G: (6:3 5:5 4:5)

F / G
```

And actually, I'm not even sure how you'd write a chord slide in marktab without this. 

So yes, modifiers should work on sequences and chord definitions.

**Comment by cknadler** [comment]
Modifiers, not transitions. Also...sequences, not chords.

For instance:

```
Riff1: [1:2 (6:1 2:1) 1:4]

Riff1*
```

**Comment by cknadler** [comment]
I do allow for transitions on chords and sequences in my implementation.

**Comment by russplaysguitar** [comment]
Yeah, but I was talking about chords and transitions, too. 

I think modifiers on sequences are good. Such as `[1:1 2 3]x` to play a sequence muted.

**Comment by cknadler** [comment]
How would modifiers work for a sequence? 

`x` is the string mute, which should error out in that case. But, the palm mute modifier `m`, in place of `x` in that same scenario, `[1:1 2 3]m`, how would that work? 
- Would it apply the modifier to every note and chord and sub-sequence, in that sequence? 
- What if some of the notes already have modifiers, would they be overwritten or left as they are?
- What if there are nested sequences with modifiers, would the outermost sequence modifier take priority or would each inner sequence apply its own modifier?

I definitely think it would be cool, I just cant figure out exactly how it would work.

**Comment by russplaysguitar** [comment]
Aw snap, good questions. (I definitely meant palm mute, not string mute.)

"Would it apply the modifier to every note and chord and sub-sequence, in that sequence?"
- Yes. That is the most straightforward way for it to work. 

"What if some of the notes already have modifiers, would they be overwritten or left as they are?"
- Overwritten. (Imagine you have a muted sequence that you want to re-use with harmonics instead.)

"What if there are nested sequences with modifiers, would the outermost sequence modifier take priority or would each inner sequence apply its own modifier?"
- Outermost should take priority. 

Basically, it has to modify everything in the sequence otherwise the rules are going to get complicated and unmaintainable. 

Logically, if you wrap stuff into a container and then express that it should be modified, it makes sense for everything in the container to be modified. 

I mostly see this feature being used to turn sequences into palm-muted sequences, since harmonics don't occur on every fret, and bending every note in a sequence is just weird. I guess a sequence could be made "more epic" by adding vibrato to every note, too.

**Comment by cknadler** [comment]
Agree with every word. Also, how can you play a note on the guitar without vibrato? I just shake my hands like crazy, every note I play, epic as possible, live it.

**Comment by russplaysguitar** [comment]
Haha, cool. This is definitely an interesting feature. It should work for chords too, right?

**Comment by cknadler** [comment]
Completely. I actually dont allow individual notes to be modified in a chord definition or inline chord production.

For example:

```
(1:5* 2:1)
```

would actually error out for me. Should I be allowing for modifiers in chords?

As far as applying them to chords, I'm definitely allowing that:

```
(1:5 2:1)*
```

is totally legitimate.

**Comment by russplaysguitar** [comment]
"Should I be allowing for modifiers in chords?"
- I can't think of a case when this is valid.

**Comment by cknadler** [comment]
- Playing a chord with a partial palm mute
- Playing a chord with some strings that are harmonics

**Comment by russplaysguitar** [comment]
Are there any songs that do that? Those seem like such edge cases...

**Comment by cknadler** [comment]
They definitely are. I only know of one of the top of my head with partial palm muting (where its actually important). Lets leave it out of the first version of the grammar.

**Comment by cknadler** [comment]
Ok, feel like we nailed this down. What I'm going to start doing is writing tests to describe the results of these discussions as well as making notes in the grammar. Feel free to beat me to either. :smile_cat: 

I'll leave this issue open until both are done.
